examples/Superresolution/enet-pat.py

```diff
@@ -0,0 +1,330 @@
+
+        if not self.inference:
+            with tf.variable_scope('discrim'):
+                real_score = discriminator(real_hr)
+                fake_score = discriminator(fake_hr)
+
+            self.build_losses(real_score, fake_score)
+
+            if ctx.is_training:
+                additional_losses = additional_losses(fake_hr, real_hr)
+
+                with tf.name_scope('additional_losses'):
+                    # see table 2 from appendix
+                    loss = []
+                    loss.append(tf.multiply(1., self.g_loss, name="loss_LA"))
+                    loss.append(tf.multiply(2e-1, additional_losses[0], name="loss_LP1"))
+                    loss.append(tf.multiply(2e-2, additional_losses[1], name="loss_LP2"))
+                    loss.append(tf.multiply(3e-7, additional_losses[2], name="loss_LT1"))
+                    loss.append(tf.multiply(1e-6, additional_losses[3], name="loss_LT2"))
+                    loss.append(tf.multiply(1e-6, additional_losses[4], name="loss_LT3"))
+
+                self.g_loss = self.g_loss + tf.add_n(loss, name='total_g_loss')
+                add_moving_summary(self.g_loss, *loss)
+
+            fake_hr = add_VGGMeans(fake_hr)
+            real_hr = add_VGGMeans(real_hr)
+            Ibicubic = add_VGGMeans(Ibicubic)
+
+            if ctx.is_training:
+                # visualization
+                viz = (tf.concat([Ibicubic, fake_hr, real_hr], 2)) * 255.
+                viz = tf.cast(tf.clip_by_value(viz, 0, 255), tf.uint8, name='viz')
+                tf.summary.image('input,fake,real', viz,
+                                 max_outputs=max(30, BATCH_SIZE))
+
+            self.collect_variables()
+
+    def _get_optimizer(self):
+        lr = tf.get_variable(
+            'learning_rate', initializer=1e-4, trainable=False)
+        opt = tf.train.AdamOptimizer(lr)
+        gradprocs = [gradproc.ScaleGradient([('VGG19/*', 0.)]), gradproc.ScaleGradient([('discrim/*', 0.3)])]
+        return optimizer.apply_grad_processors(opt, gradprocs)
+
+
+def apply(model_path, lowres_path="", output_path=None):
+    assert os.path.isfile(lowres_path)
+    assert output_path is not None
+    lr = cv2.imread(lowres_path).astype(np.float32)
+    baseline = cv2.resize(lr, (0, 0), fx=4, fy=4, interpolation=cv2.INTER_CUBIC)
+    LR_SIZE_H, LR_SIZE_W = lr.shape[:2]
+    lr[:, :, 0] -= VGG_MEAN[0]
+    lr[:, :, 1] -= VGG_MEAN[1]
+    lr[:, :, 2] -= VGG_MEAN[2]
+
+    predict_func = OfflinePredictor(PredictConfig(
+        model=Model(LR_SIZE_H, LR_SIZE_W, inference=True),
+        session_init=get_model_loader(model_path),
+        input_names=['Ilr'],
+        output_names=['prediction']))
+
+    pred = predict_func(lr[None, ...])
+    p = np.clip(pred[0][0, ...] * 255, 0, 255)
+
+    cv2.imwrite(output_path + "predition.png", p)
+    cv2.imwrite(output_path + "baseline.png", baseline)
+
+
+def get_data(lmdb):
+    ds = LMDBDataPoint(lmdb, shuffle=True)
```

Comment:
Is this faster than putting `ImageDataFromZIPFile` + preprocess here?
I guess maybe just the `zipfile` module is really slow. You are doing random read with LMDB anyway.

Reply:
At some point I did benchmarks.